### ISSUE TITLE
WMStats CouchDB database compaction never converging

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#10779]
**Impact of the new feature**
CouchDB - WMStats db 

**Is your feature request related to a problem? Please describe.**
This is actually not a new issue, and it's the reason why from time to time (once or twice a year) we have to rotate the wmstats database (basically dumping the database and letting the agents replicate data back to central couch).

Said that, we have been struggling to compact the wmstats database for the last 2 or 3 weeks. The compaction process never catches up to the latest document updates, once it finishes what it was compacting, it finds another few 10s of thousands of documents to compact, when it's over, it cannot swap the database because again there are document updates to be taken into consideration, and so on.
This process usually goes through fairly quickly, but I think the database is very very large at this moment, thus having a performance degradation and becoming slower than the speed the agents write data to it.

There were times, over the last weeks, that the database went up to +300GB of data volume, when it's supposed to be between 30GB and 100GB. The only way I found to allow the last step of the compaction to finish (db swap), was by stopping all the agents (components only) that had thousands of jobs in their queue, thus stopping to write docs to the database. Once compaction was completed, all the agents were restarted.

**Describe the solution you'd like**
Immediate solution will be the wmstats database rotation. In the past, we had simply dumped the wmstats database, restarted couchdb in the appropriate backend, and let it get recreated from the content available in the agents db. Another option would be to replicate valid documents to another database, and swap then once the replication is over (this can be in parallel with the system; or with an afternoon of timeout for central wmstats).

Long term solution is: evaluate what exactly we write to the wmstats database, and how we write to it (via replication or via http requests), and try to redesign it to make it lighter and with a better performance. One of my ideas is to stop making document updates for every single job that changes its WMBS state (replace it by a summary provided by each agent).

**Describe alternatives you've considered**
Need to decide whether we a) drop the database or b) create a service downtime, replicate non-deleted docs, and swap dbs when ready.

**Additional context**
We had almost 1B updates made in this database, it currently has 5M docs and 18M deleted docs:
```
[cmsweb@vocms0743 ~]$ curl http://localhost:5984/wmstats
{"db_name":"wmstats","doc_count":4698639,"doc_del_count":18518575,"update_seq":975752940,"purge_seq":0,"compact_running":true,"disk_size":262137974919,"data_size":19054093262,"instance_start_time":"1629484349120196","disk_format_version":6,"committed_update_seq":975752940}
```

**Comment by amaltaro** [comment]
wmstats db rotation seems to be back to normal behaviour, now that many of the old agents have been finally shut off (only 1 remaining). So, we might keep wmstats database like is, or rotate it in the coming week, once Hasan manages to go through workflows in need of assistance.

**Comment by amaltaro** [comment]
The `wmstats` database was dropped during the CouchDB upgrade and I think this issue can be closed out now. Let's see how compaction behaves in the new CouchDB 3.1.2 version.
